Beam/Include/Beam/Threading/Sync.hpp

```diff
@@ -27,6 +61,50 @@ namespace Threading {
       //! The type of mutex used.
       using Mutex = MutexType;
 
+      //! The type of the lock used for immutable access.
+      using ReadLock = Details::GetReadLock<Mutex>;
+
+      //! The type of the lock used for mutable access.
+      using WriteLock = Details::GetWriteLock<Mutex>;
+
+      //! The proxy for any type of lock in use.
+      class LockProxy {
+        public:
+          //! Construct an empty proxy.
+          LockProxy() = default;
+
+          //! Constructs a proxy to a lock.
+          /*!
+            \tparam The type of the lock.
+            \param lock The pointer to the lock.
+          */
+          template<typename T>
+          explicit LockProxy(T* lock)
+            : m_lock(lock) {}
+      
+          //! Locks the lock referenced.
+          void lock() {
+            std::visit(
+              [] (auto& ptr) {
```

Comment:
How is this compiling? ```std::visit``` needs two arguments.

Reply:
It is not. This family of functions (`Release`, `GetLock`, etc.) do not seem to be complied into any code, and the compiler is unable to detect this during the first lookup, so such stupid bugs are quite easy to miss. I will fix it.

Reply:
Okay, what you can do is try to compile Nexus using this branch.

Go to wherever you checked out Nexus, then go to the Dependencies/Beam folder, checkout this branch and try compiling Nexus.